src/vos/vos_obj_cache.c

```diff
@@ -238,69 +287,134 @@ vos_obj_discard_hold(struct daos_lru_cache *occ, struct vos_container *cont, dao
+	obj_old->obj_ih         = DAOS_HDL_INVAL;
+	clean_object(obj_old);
+	switch (action) {
+	default:
+		D_ASSERTF(0, "Invalid action: %d\n", action);
+	case VOS_OBJ_EPH_NONE:
+		memset(obj_old, 0, sizeof(*obj_old));
+		break;
+	case VOS_OBJ_EPH_DROP:
+		obj_old->obj_dropped = 1;
+		break;
+	case VOS_OBJ_EPH_FREE:
+		D_FREE(obj_old);
+	}
+}
+
+/** Move an object to the lru cache */
 static inline int
-cache_object(struct daos_lru_cache *occ, struct vos_object **objp)
+cache_object(struct vos_obj_cache *occ, bool ephemeral, struct vos_object *obj_old,
+	     struct vos_object **objp)
 {
-	struct vos_object	*obj_new;
+	struct vos_object       *obj_new = NULL;
 	struct daos_llink	*lret;
 	struct obj_lru_key	 lkey;
-	int			 rc;
+	struct vos_object       *old_eph = NULL;
+	int                      rc      = 0;
+	int                      action  = VOS_OBJ_EPH_NONE;
 
 	*objp = NULL;
 
-	D_ASSERT(obj_local.obj_cont != NULL);
+	D_ASSERT(obj_old->obj_cont != NULL);
 
-	lkey.olk_cont = obj_local.obj_cont;
-	lkey.olk_oid = obj_local.obj_id;
+	if (obj_old->obj_ephemeral) {
+		/* If it's in use, we are taking it over, otherwise, we are
+		 * replacing it and need to free the old entry
+		 */
+		action = occ->oc_eph_in_use ? VOS_OBJ_EPH_DROP : VOS_OBJ_EPH_FREE;
+	} else if (occ->oc_eph_cont != NULL) {
+		old_eph = occ->oc_ephemeral; /* If we replace, we will need to clean this one */
+	}
+
+	if (ephemeral) {
+		D_ALLOC_PTR(obj_new);
+		if (obj_new == NULL)
+			rc = -DER_NOMEM;
+		else
+			init_object(obj_new, obj_old->obj_id, obj_old->obj_cont);
+	} else {
+		lkey.olk_cont = obj_old->obj_cont;
+		lkey.olk_oid  = obj_old->obj_id;
+
+		rc = daos_lru_ref_hold(occ->oc_lru, &lkey, sizeof(lkey), obj_old->obj_cont, &lret);
+		if (rc == 0) {
+			/** Object is in cache */
+			obj_new = container_of(lret, struct vos_object, obj_llink);
+		}
+	}
 
-	rc = daos_lru_ref_hold(occ, &lkey, sizeof(lkey), obj_local.obj_cont, &lret);
 	if (rc != 0) {
-		clean_object(&obj_local);
-		memset(&obj_local, 0, sizeof(obj_local));
-		return rc; /* Can't cache new object */
+		if (!obj_old->obj_ephemeral) {
+			clean_object(obj_old);
+			memset(obj_old, 0, sizeof(*obj_old));
+		}
+		return rc;
 	}
 
-	/** Object is in cache */
-	obj_new = container_of(lret, struct vos_object, obj_llink);
-	/* This object should not be cached */
-	D_ASSERT(obj_new->obj_df == NULL);
+	move_object(obj_new, obj_old, action);
 
-	vos_ilog_fetch_move(&obj_new->obj_ilog_info, &obj_local.obj_ilog_info);
-	obj_new->obj_toh = obj_local.obj_toh;
-	obj_new->obj_ih = obj_local.obj_ih;
-	obj_new->obj_sync_epoch = obj_local.obj_sync_epoch;
-	obj_new->obj_df = obj_local.obj_df;
-	obj_new->obj_zombie = obj_local.obj_zombie;
-	obj_local.obj_toh = DAOS_HDL_INVAL;
-	obj_local.obj_ih = DAOS_HDL_INVAL;
-	clean_object(&obj_local);
-	memset(&obj_local, 0, sizeof(obj_local));
+	if (ephemeral) {
+		if (old_eph) {
+			clean_object(old_eph);
+			if (occ->oc_eph_in_use)
+				old_eph->obj_dropped = 1;
+			else
+				D_FREE(old_eph);
+		}
```

Comment:
This looks duplicated with the latter part (start from switch(action)) in move_object(), it would be more read-able if you move that part out of move_object().